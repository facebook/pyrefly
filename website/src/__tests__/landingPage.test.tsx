/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import Home from '../pages/index';

describe('Home page', () => {
    test('renders internal site correctly', () => {
        process.env.INTERNAL_STATIC_DOCS = '1';
        const { container } = render(<Home />);

        expectHeaderAndTitleIsCorrect(container);

        // Verify that the new sections are present
        const whyPyreFlySection = document.getElementById(
            'why-pyrefly-section'
        );
        expect(whyPyreFlySection).toBeInTheDocument();

        const performanceComparisonSection = document.getElementById(
            'performance-comparison-section'
        );
        expect(performanceComparisonSection).toBeInTheDocument();

        // Run test with --update-snapshot to update the snapshot if the test is failing after
        // you made a intentional change to the home page
        expect(container).toMatchSnapshot();
    });

    test('renders external site correctly', () => {
        process.env.INTERNAL_STATIC_DOCS = undefined;
        const { container } = render(<Home />);

        expectHeaderAndTitleIsCorrect(container);

        // Verify that the new sections are not present
        const whyPyreFlySection = document.getElementById(
            'why-pyrefly-section'
        );
        expect(whyPyreFlySection).toBeInTheDocument();

        const performanceComparisonSection = document.getElementById(
            'performance-comparison-section'
        );
        expect(performanceComparisonSection).toBeInTheDocument();

        const quotesSection = document.getElementById('quotes-section');
        expect(quotesSection).not.toBeInTheDocument();

        // Run test with --update-snapshot to update the snapshot if the test is failing after
        // you made a intentional change to the home page
        expect(container).toMatchSnapshot();
    });

    function expectHeaderAndTitleIsCorrect(container: HTMLElement) {
        // Verify that title is set correctly
        expect(document.title).toBe(
            'Pyrefly: A Fast Python Type Checker and Language Server'
        );

        // Verify that fireflies are present
        const fireflyElements = document.querySelectorAll('#firefly');
        expect(fireflyElements.length).toBeGreaterThan(1);

        // Verify that header is correct
        const header = container.querySelector('header');
        expect(header).toBeInTheDocument();

        // Verify that the logo image is present
        const logoImg = header?.querySelector(
            'img[src="/img/Pyrefly-Brandmark.svg"]'
        );
        expect(logoImg).toBeInTheDocument();
        expect(logoImg?.getAttribute('alt')).toBe('Pyrefly Logo');

        // verify subtitle is correct
        const subtitle = container.querySelector('header p');
        expect(subtitle).toBeInTheDocument();
        const subtitleText = subtitle?.textContent?.trim();
        expect(subtitleText).toBe(
            'A fast type checker and language server for Python with powerful IDE features'
        );
    }
});
