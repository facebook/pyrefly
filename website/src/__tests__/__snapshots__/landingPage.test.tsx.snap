// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders external site correctly 1`] = `
<div>
  <div>
    <div>
      Pyrefly: A Static Type Checker for Python
    </div>
    <div>
      Pyrefly is a static type checker for Python
    </div>
    <div>
      <section
        id="header-section"
      >
        <div
          class="container"
        >
          <header>
            <div
              style="opacity: 0;"
            >
              <section>
                <img
                  alt="Pyrefly Logo"
                  src="/img/Pyrefly-Brandmark.svg"
                />
              </section>
            </div>
            <div
              style="opacity: 0;"
            >
              <p>
                <span>
                  A faster Python type checker written in Rust
                </span>
              </p>
            </div>
            <div
              style="opacity: 0;"
            >
              <section>
                <div>
                  <pre>
                    <code>
                      <span>
                        $ 
                      </span>
                      pip install pyrefly && pyrefly init
                    </code>
                    <button
                      title="Copy to clipboard"
                    >
                      <svg
                        height="20"
                        viewBox="340 364 14 15"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M342 375.974h4v.998h-4v-.998zm5-5.987h-5v.998h5v-.998zm2 2.994v-1.995l-3 2.993 3 2.994v-1.996h5v-1.995h-5zm-4.5-.997H342v.998h2.5v-.997zm-2.5 2.993h2.5v-.998H342v.998zm9 .998h1v1.996c-.016.28-.11.514-.297.702-.187.187-.422.28-.703.296h-10c-.547 0-1-.452-1-.998v-10.976c0-.546.453-.998 1-.998h3c0-1.107.89-1.996 2-1.996 1.11 0 2 .89 2 1.996h3c.547 0 1 .452 1 .998v4.99h-1v-2.995h-10v8.98h10v-1.996zm-9-7.983h8c0-.544-.453-.996-1-.996h-1c-.547 0-1-.453-1-.998 0-.546-.453-.998-1-.998-.547 0-1 .452-1 .998 0 .545-.453.998-1 .998h-1c-.547 0-1 .452-1 .997z"
                          fill="currentColor"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </pre>
                </div>
                <a
                  href="https://marketplace.visualstudio.com/items?itemName=meta.pyrefly"
                  target="_blank"
                >
                   
                  Get VSCode Extension
                   
                </a>
              </section>
            </div>
            <section>
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
            </section>
          </header>
        </div>
      </section>
      <section
        id="why-pyrefly-section"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <div>
              <div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0s;"
                  >
                    <h3>
                      Scale with Confidence
                    </h3>
                    <p>
                      Type check over 1.85 million lines of code
                       
                      <span>
                        per second.
                        <sup>
                          <span>
                            <span>
                              ⓘ
                            </span>
                            <span>
                              Tested using Meta infrastructure (166 cores, 228 GB RAM)
                            </span>
                          </span>
                        </sup>
                      </span>
                    </p>
                  </div>
                </div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0.05s;"
                  >
                    <h3>
                      Developer Delight
                    </h3>
                    <p>
                      Get lightning fast autocomplete, and catch errors with instant feedback in your favorite editor.
                    </p>
                  </div>
                </div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0.1s;"
                  >
                    <h3>
                      Easy Onboarding
                    </h3>
                    <p>
                      <a
                        href="/en/docs/installation/"
                        target="_blank"
                      >
                        Start type checking
                      </a>
                       your code in minutes.
                    </p>
                  </div>
                </div>
              </div>
              <section>
                <li
                  id="firefly"
                />
                <li
                  id="firefly"
                />
                <li
                  id="firefly"
                />
              </section>
            </div>
          </div>
        </div>
      </section>
      <section
        id="performance-comparison-section"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <h2>
              Performance Comparison
            </h2>
          </div>
          <div
            style="opacity: 0;"
          >
            <div>
              <div>
                <button>
                  PyTorch
                </button>
                <button>
                  Instagram
                </button>
              </div>
              <div>
                <div>
                  Type checking the
                   
                  <a
                    href="https://github.com/pytorch/pytorch"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    PyTorch
                  </a>
                   
                  codebase from scratch.
                  <span>
                    <span>
                      ⓘ
                    </span>
                    <span>
                      <strong>
                        Tested using Macbook
                      </strong>
                      <br />
                      (10 cores: 8 performance + 2 efficiency cores, 32 GB RAM)
                    </span>
                  </span>
                </div>
                <div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            Pyrefly
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "pyrefly check"
                            </strong>
                            <br />
                            Pyrefly uses as many threads as possible
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 4.827299209321681%;"
                        >
                          <div
                            style="animation-duration: 2.32s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            Pyright
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "pyright --threads=8"
                            </strong>
                            <br />
                            8 threads yielded the best performance after testing multiple settings
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 73.1585518102372%;"
                        >
                          <div
                            style="animation-duration: 35.16s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            MyPy
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "dmypy run"
                            </strong>
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 100%;"
                        >
                          <div
                            style="animation-duration: 48.06s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        id="pyrefly-video"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <h2>
              Check it out in the intro video
            </h2>
          </div>
          <div
            style="opacity: 0;"
          >
            <div>
              <link
                as="image"
                href="https://i.ytimg.com/vi/LXaFRKrTJVU/maxresdefault.jpg"
                rel="preload"
              />
              <article
                class="yt-lite "
                data-title="Introducing Pyrefly: A new type checker and IDE experience for Python"
                style="background-image: url(https://i.ytimg.com/vi/LXaFRKrTJVU/maxresdefault.jpg); --aspect-ratio: 56.25%;"
              >
                <button
                  aria-label="Watch Introducing Pyrefly: A new type checker and IDE experience for Python"
                  class="lty-playbtn"
                  type="button"
                />
              </article>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`Home page renders internal site correctly 1`] = `
<div>
  <div>
    <div>
      Pyrefly: A Static Type Checker for Python
    </div>
    <div>
      Pyrefly is a static type checker for Python
    </div>
    <div>
      <section
        id="header-section"
      >
        <div
          class="container"
        >
          <header>
            <div
              style="opacity: 0;"
            >
              <section>
                <img
                  alt="Pyrefly Logo"
                  src="/img/Pyrefly-Brandmark.svg"
                />
              </section>
            </div>
            <div
              style="opacity: 0;"
            >
              <p>
                <span>
                  A faster Python type checker written in Rust
                </span>
              </p>
            </div>
            <div
              style="opacity: 0;"
            >
              <section>
                <div>
                  <pre>
                    <code>
                      <span>
                        $ 
                      </span>
                      pip install pyrefly && pyrefly init
                    </code>
                    <button
                      title="Copy to clipboard"
                    >
                      <svg
                        height="20"
                        viewBox="340 364 14 15"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M342 375.974h4v.998h-4v-.998zm5-5.987h-5v.998h5v-.998zm2 2.994v-1.995l-3 2.993 3 2.994v-1.996h5v-1.995h-5zm-4.5-.997H342v.998h2.5v-.997zm-2.5 2.993h2.5v-.998H342v.998zm9 .998h1v1.996c-.016.28-.11.514-.297.702-.187.187-.422.28-.703.296h-10c-.547 0-1-.452-1-.998v-10.976c0-.546.453-.998 1-.998h3c0-1.107.89-1.996 2-1.996 1.11 0 2 .89 2 1.996h3c.547 0 1 .452 1 .998v4.99h-1v-2.995h-10v8.98h10v-1.996zm-9-7.983h8c0-.544-.453-.996-1-.996h-1c-.547 0-1-.453-1-.998 0-.546-.453-.998-1-.998-.547 0-1 .452-1 .998 0 .545-.453.998-1 .998h-1c-.547 0-1 .452-1 .997z"
                          fill="currentColor"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </pre>
                </div>
                <a
                  href="https://marketplace.visualstudio.com/items?itemName=meta.pyrefly"
                  target="_blank"
                >
                   
                  Get VSCode Extension
                   
                </a>
              </section>
            </div>
            <section>
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
              <li
                id="firefly"
              />
            </section>
          </header>
        </div>
      </section>
      <section
        id="why-pyrefly-section"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <div>
              <div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0s;"
                  >
                    <h3>
                      Scale with Confidence
                    </h3>
                    <p>
                      Type check over 1.85 million lines of code
                       
                      <span>
                        per second.
                        <sup>
                          <span>
                            <span>
                              ⓘ
                            </span>
                            <span>
                              Tested using Meta infrastructure (166 cores, 228 GB RAM)
                            </span>
                          </span>
                        </sup>
                      </span>
                    </p>
                  </div>
                </div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0.05s;"
                  >
                    <h3>
                      Developer Delight
                    </h3>
                    <p>
                      Get lightning fast autocomplete, and catch errors with instant feedback in your favorite editor.
                    </p>
                  </div>
                </div>
                <div
                  style="opacity: 0;"
                >
                  <div
                    style="transition-delay: 0.1s;"
                  >
                    <h3>
                      Easy Onboarding
                    </h3>
                    <p>
                      <a
                        href="/en/docs/installation/"
                        target="_blank"
                      >
                        Start type checking
                      </a>
                       your code in minutes.
                    </p>
                  </div>
                </div>
              </div>
              <section>
                <li
                  id="firefly"
                />
                <li
                  id="firefly"
                />
                <li
                  id="firefly"
                />
              </section>
            </div>
          </div>
        </div>
      </section>
      <section
        id="performance-comparison-section"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <h2>
              Performance Comparison
            </h2>
          </div>
          <div
            style="opacity: 0;"
          >
            <div>
              <div>
                <button>
                  PyTorch
                </button>
                <button>
                  Instagram
                </button>
              </div>
              <div>
                <div>
                  Type checking the
                   
                  <a
                    href="https://github.com/pytorch/pytorch"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    PyTorch
                  </a>
                   
                  codebase from scratch.
                  <span>
                    <span>
                      ⓘ
                    </span>
                    <span>
                      <strong>
                        Tested using Macbook
                      </strong>
                      <br />
                      (10 cores: 8 performance + 2 efficiency cores, 32 GB RAM)
                    </span>
                  </span>
                </div>
                <div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            Pyrefly
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "pyrefly check"
                            </strong>
                            <br />
                            Pyrefly uses as many threads as possible
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 4.827299209321681%;"
                        >
                          <div
                            style="animation-duration: 2.32s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            Pyright
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "pyright --threads=8"
                            </strong>
                            <br />
                            8 threads yielded the best performance after testing multiple settings
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 73.1585518102372%;"
                        >
                          <div
                            style="animation-duration: 35.16s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                  <div>
                    <div>
                      <div>
                        <span>
                          <strong>
                            MyPy
                          </strong>
                        </span>
                        <span>
                          <span>
                            ⓘ
                          </span>
                          <span>
                            <strong>
                              Command: "dmypy run"
                            </strong>
                          </span>
                        </span>
                      </div>
                    </div>
                    <div>
                      <div>
                        <div
                          style="height: 80%; width: 100%;"
                        >
                          <div
                            style="animation-duration: 48.06s; width: 0%;"
                          />
                        </div>
                      </div>
                      <span>
                        <strong>
                          0.0s
                        </strong>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        id="pyrefly-video"
      >
        <div
          class="container"
        >
          <div
            style="opacity: 0;"
          >
            <h2>
              Check it out in the intro video
            </h2>
          </div>
          <div
            style="opacity: 0;"
          >
            <div>
              <link
                as="image"
                href="https://i.ytimg.com/vi/LXaFRKrTJVU/maxresdefault.jpg"
                rel="preload"
              />
              <article
                class="yt-lite "
                data-title="Introducing Pyrefly: A new type checker and IDE experience for Python"
                style="background-image: url(https://i.ytimg.com/vi/LXaFRKrTJVU/maxresdefault.jpg); --aspect-ratio: 56.25%;"
              >
                <button
                  aria-label="Watch Introducing Pyrefly: A new type checker and IDE experience for Python"
                  class="lty-playbtn"
                  type="button"
                />
              </article>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
